File: java/binaryTreeMaximumPathSum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    //keep track of the max we've found so far outside of the recursive function
    int maxPath;
    
    public int maxPathSum(TreeNode root) {
        maxPath = root.val;
        getMaxPath(root);
        return maxPath;
    }
    
    private int getMaxPath(TreeNode root) {
        //recursive base case
        if (root.left == null && root.right == null) {
            maxPath = Math.max(maxPath, root.val);
        }
        //we need to compare four possibilities: the v-shaped path through this root, 
        //the root by itself, the left path and the right path
        int leftPath = root.left == null ? 0 : getMaxPath(root.left);
        int rightPath = root.right == null ? 0 : getMaxPath(root.right);
        //the v path is not returned since it's not accessible to the parent roots
        int returnMax = Math.max(root.val, Math.max(root.val + leftPath, root.val + rightPath));
        maxPath = Math.max(maxPath, Math.max(returnMax, leftPath + rightPath + root.val));
        return returnMax;
    }
}

File: python/binary_tree_maximum_path_sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: TreeNode) -> int:
        #keep track of the max we've found so far outside of the recursive function
        self.maxPath = root.val
        self.getMaxPath(root)
        return self.maxPath
        
    def getMaxPath(self, root: TreeNode) -> int:
        #recursive base case
        if root.left is None and root.right is None:
            self.maxPath = max(self.maxPath, root.val)
            return root.val
        #we need to compare four possibilities: the v-shaped path through this root, 
        # the root by itself, the left path and the right path
        leftMaxPath = self.getMaxPath(root.left) if root.left is not None else 0
        rightMaxPath = self.getMaxPath(root.right) if root.right is not None else 0
        vPath = root.val + leftMaxPath + rightMaxPath
        leftPath = leftMaxPath + root.val
        rightPath = rightMaxPath + root.val
        self.maxPath = max(self.maxPath, root.val, leftPath, rightPath, vPath)
        #the v path is not returned since it's not accessible to the parent roots
        return max(root.val, leftPath, rightPath)